### ISSUE TITLE
Unathorized use of silero-models

### ISSUE THREAD

**Issue by snakers4** [natsuite/NatML-Py#1]
Hi,

My name is Alexander Veysov, I am one of authors of silero-models - https://github.com/snakers4/silero-models.

We did not authorize your reverse engineering, re-hosting and usage of our products in your, among other, private products under different license.

Also we did not authorize re-posting our models into commericial asset stores.

Please remove these immediately.

Best,
Alexander

Links:
 
- https://assetstore.unity.com/packages/tools/integration/silero-stt-realtime-speech-to-text-201282
- https://hub.natml.ai/@natsuite/silero-stt-v3

**Comment by olokobayusuf** [comment]
> We did not authorize your reverse engineering

There is no reverse-engineering of the model or source code. The model is used as-is, and the Unity code is implemented from scratch in C# to be able to run in Unity's scripting environment.

> re-hosting and usage of our products in your, among other, private products under different license.

The [package on the Unity asset store](https://assetstore.unity.com/packages/tools/integration/silero-stt-realtime-speech-to-text-201282) is distributed under the GNU Affero GPL v3 license, same as your original repository:

<img width="803" alt="Screen Shot 2021-11-06 at 10 51 34 AM" src="https://user-images.githubusercontent.com/17663473/140613854-95b46705-589b-4d19-91b6-c65c5a111023.png">

> Also we did not authorize re-posting our models into commericial asset stores.

The AGPL v3 license allows for re-distirbution, compliant with the original license.

> Please remove these immediately.

I'll go ahead and remove the model from NatML Hub and the Unity store, but it seems that you didn't actually intend to distribute the models under AGPLv3 to begin with. I strongly recommend that you update your upstream license to something that only allows you--and no one else--to distribute it.

**Comment by snakers4** [comment]
> The package on the Unity asset store is distributed under the GNU Affero GPL v3 license, same as your original repository:

As far as I see, the packages there are published under some other license called "Asset Store Terms of Service and EULA".

![image](https://user-images.githubusercontent.com/12515440/140616157-80d3a344-9ac2-45ec-a7bd-691ec761b1fc.png)

So basically what you are doing is you are populating some corporations' asset store with intellectual properly taken from dozens or hundreds of authors for your profit without even having a decency of contacting them or sharing the earned revenue. In the meantime, the license gets "purified". This is more or less akin to Facebook copying videos from YouTube.

I have created a ticket with their support regarding these numerous re-uploads to clarify this issue further, because sharing / marketing the models for non-profit use is one thing (customary in the ML community), but actually selling broken products at other people's expense is another (judging by the comments, your re-uploads did not also work, which casts a shadow on us as the authors).

> I'll go ahead and remove the model from NatML Hub and the Unity store, but it seems that you didn't actually intend to distribute the models under AGPLv3 to begin with. I strongly recommend that you update your upstream license to something that only allows you--and no one else--to distribute it.

We chose AGPLv3 because our repo is basically a demo project and most corporations are allergic to this license. There are no real popular licenses 100% suitable for demo / non-commercial demonstration purposes.

On a more general note, I see more and more often the approach where opportunistic agents just take whatever models they can find and show them off as their own, trying to re-sell them to unsuspecting customers. This more or less is detrimental to the ML community in general, because it demotivates authors to do honest work / business and incentives such predatory behavior.

Shame on you.

**Comment by olokobayusuf** [comment]
> As far as I see, the packages there are published under some other license called "Asset Store Terms of Service and EULA".

There are two things being distributed in predictor packages: the ML model and the predictor code that makes predictions using the model. In this case, the model is published under AGPLv3, which is consistent with your repo, **but the code that makes predictions with said model does not have to be published under the same license**. Many Unity integration plugins work under this model, like the [OpenCV package](https://assetstore.unity.com/packages/tools/integration/opencv-for-unity-21088) which sells trampoline code from Unity into the open-source OpenCV model. In this case specifically, if you had downloaded the predictor package from NatML Hub instead of the Unity asset store, **you would see that even the predictor code is licensed under AGPL**.

> So basically what you are doing is you are populating some corporations' asset store with intellectual properly taken from dozens or hundreds of authors for your profit without even having a decency of contacting them or sharing the earned revenue.

Again, there are two things to unpack here: all models on NatML Hub are published under licenses that allow for both re-distribution and commercial use. This isn't some evil plot, and again, what I am selling isn't the machine learning model; it is the code that uses the ML model with Unity's execution environment. Unity has its own API's for handling things like images that are not what you would see in Python. And finally, I did reach out to you all to bring your models to Hub and I got no response.

> I have created a ticket with their support regarding these numerous re-uploads to clarify this issue further, because sharing / marketing the models for non-profit use is one thing (customary in the ML community), but actually selling broken products at other people's expense is another (judging by the comments, your re-uploads did not also work, which casts a shadow on us as the authors).

That's fine; there's nothing of substance in these tickets. Given that you are an ML researcher, you know that using ML models requires a fair amount of pre- and post-processing code, all of which must be specific to a development environment. Just to re-emphasize this, what is being sold is the ML predictor code, **not the ML model**. Anyone can download any pre-trained model from any repository on GitHub, but good luck converting anchor boxes and logits to rectangles for example.

> On a more general note, I see more and more often the approach where opportunistic agents just take whatever models they can find and show them off as their own

Again, where exactly have we shown off the model as our own? Every single page links back to the original model source, and your model is no exception.

> trying to re-sell them to unsuspecting customers

What you seem to be missing is that most developers couldn't care less about the intricacies of your machine learning model. They just want to call a single function in their code and get some usable result. This is what we do. We sell extremely simple API's built around these ML models, because literally nobody cares about the models themselves (as much as you clearly do). It's the same as anyone paying for STT services on AWS or GCP.

> This more or less is detrimental to the ML community in general, because it demotivates authors to do honest work / business and incentives such predatory behavior.

Says who? If this is so much of an issue, then the likes of [HuggingFace](https://huggingface.co/) are also being evil by selling quick API's for using the multitude of NLP models in the wild. Investors, along with thousands of businesses that deploy ML models in their products, disagree with you.

**Comment by olokobayusuf** [comment]
[Here's an example of ML researchers who understand the dynamic I describe above](https://github.com/PeterL1n/RobustVideoMatting/issues/36).

**Comment by snakers4** [comment]
> There are two things being distributed in predictor packages: the ML model and the predictor code that makes predictions using the model. In this case, the model is published under AGPLv3, which is consistent with your repo, but the code that makes predictions with said model does not have to be published under the same license. Many Unity integration plugins work under this model, like the OpenCV package which sells trampoline code from Unity into the open-source OpenCV model. In this case specifically, if you had downloaded the predictor package from NatML Hub instead of the Unity asset store, you would see that even the predictor code is licensed under AGPL.

![image](https://user-images.githubusercontent.com/12515440/140617929-5ebfec28-bb51-4514-bf9b-991932839502.png)

Interestingly, you speak about re-licensing.
Looks like you did the re-licensing on your side without our consent.

> Says who? If this is so much of an issue, then the likes of HuggingFace are also being evil by selling quick API's for using the multitude of open-source NLP models in the wild. Investors, along with thousands of businesses that deploy ML models in their products, disagree with you.

Deploying and tuning a model that is actually useful in production has very little in common with whatever these guys do. All modern NLP transformer craze also is very misguided.

> Here's an example of ML researchers who understand the dynamic I describe above.
> ByteDance Inc. | University of Washington

These people were paid by their investor / state / university to publish the model.
Their work is pre-financed. They do not care about commercializing their models. They do not invest their own money into their research.
Their public model release in the first place is a planned PR activity, so in a sense you do their PR for them.

> That's fine; there's nothing of substance in these tickets. 
>  Just to re-emphasize this, what is being sold is the ML predictor code, not the ML model. Anyone can download any pre-trained model from any repository on GitHub, but good luck converting anchor boxes and logits to rectangles for example.

This is a clever evasion, and it even may work in the corporation behind Unity does not care.
But what you are doing in a nutshell is building your capitalization at the expense of community, which will lead to research siloing at GAFA level companies.

> Again, where exactly have we shown off the model as our own? Every single page links back to the original model source, and your model is no exception.

While technically you may have a link stashed in the bottom, no one realistically will see it.
Even PyTorch, for example in their hub, gives you a place for your logo.

> What you seem to be missing is that most developers couldn't care less about the intricacies of your machine learning model. They just want to call a single function in their code and get some usable result. This is what we do. We sell extremely simple API's built around these ML models, because literally nobody cares about the models themselves (as much as you clearly do). It's the same as anyone paying for STT services on AWS or GCP.

There is a new ML winter around the corner as always, but whatever people like you are doing - actually makes ML research possible only within GAFA-like companies marginalizing the rest.

Not caring about people who make all of these models is highly cynical and unethical. 
This is a grey zone, but I always wondered how the thought process of such people worked.

I understand that 95% of models are either semi-abandoned academic artefacts or PR projects of companies like Bytedance, but not all of them.

> And finally, I did reach out to you all to bring your models to Hub and I got no response.

We read your email, it was of no interest - because it has no clear benefits for us as developers.

Essentially you want to mooch off our research to build your capitalization and give nothing in return.
Now you got your response. If someone does not reply, it does not mean he / she agrees.

If you want to build relations with people actually powering your project you should care about them. Selling their work and giving nothing in return is not a sustainable strategy. Once the big companies stop publishing models for some reason - you are no longer viable, because you do not create new models.

**Comment by olokobayusuf** [comment]
> Interestingly, you speak about re-licensing.
Looks like you did the re-licensing on your side without our consent.

Your project is licensed under AGPLv3; the predictor package on NatML Hub and the Unity Store were also licensed under AGPLv3. Where exactly did I re-license it?

> Deploying and tuning a model that is actually useful in production has very little in common with whatever these guys do. All modern NLP transformer craze also is very misguided.

Sounds like you have larger grudges with the present state of applied machine learning.

> But what you are doing in a nutshell is building your capitalization at the expense of community, which will lead to research siloing at GAFA level companies.

Again, all models on Hub are distributed under MIT or Apache 2.0. Writing paid predictor packages for them is consistent with their licenses.

> While technically you may have a link stashed in the bottom, no one realistically will see it.

This is actually great design feedback! I'll forward this to our web designers.

> Not caring about people who make all of these models is highly cynical and unethical.

I did invite you to re-license your models for sale on NatML Hub, but you ignored it. Now, we are moving away from having paid models entirely, so this point is now moot.

> Once the big companies stop publishing models for some reason - you are no longer viable, because you do not create new models.

We're actually researchers ourselves. In any case, I've unlisted your model so we can finalize this conversation.
